cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputCommitter.java

```diff
@@ -135,11 +139,28 @@ public void commitJob(JobContext context) throws IOException {
       ConfigurationUtil.getNamedConfigurations(this.taskContext.getConfiguration(),
                                                PartitionedFileSetArguments.OUTPUT_PARTITION_METADATA_PREFIX);
 
+    boolean allowAppend = partitionWriteOption == DynamicPartitioner.PartitionWriteOption.ALLOW_APPEND;
     // create all the necessary partitions
-    for (PartitionKey partitionKey : partitionsToAdd) {
-      PartitionOutput partitionOutput = outputDataset.getPartitionOutput(partitionKey);
-      partitionOutput.setMetadata(metadata);
-      partitionOutput.addPartition();
+    switch (partitionWriteOption) {
+      case ALLOW_APPEND:
+      case NEW_ONLY:
+        for (Map.Entry<String, PartitionKey> entry : partitionsToAdd.entrySet()) {
+          outputDataset.addPartition(entry.getValue(), entry.getKey(), metadata, true, allowAppend);
+        }
+        break;
+      case OVERWRITE:
+        for (Map.Entry<String, PartitionKey> entry : partitionsToAdd.entrySet()) {
+          if (outputDataset.getPartition(entry.getValue()) != null) {
+            // this allows reinstating the existing files if there's a rollback.
```

Comment:
So overwriting a partition also removes all existing metadata? Is that clear from the Javadocs? If I am not mistaken, that states that only the contents are overwritten.